### ISSUE TITLE
GET followers / following API offset ID does not handle scenario where the offset ID is unfollowed.

### ISSUE THREAD

**Issue by rmisio** [OpenBazaar/openbazaar-go#90]
GET followers / following API offset ID does not handle scenario where the guid represented by the offset ID is un-followed, or where that guid is un-followed and then re-followed (resulting in it's position moving to the top of the list).

Scenario 1:
- fetch a page of followers. for example, the last follower in the page has an guid of 'abc'.
- before you fetch the next page, user `abc` un-follows you.
- fetch the second page using 'abc' as the offsetId.
- haven't test this, but based on our discussion, I think the API may bomb or return nothing because the guid referenced by the offsetID is no longer in the list.

Expected results:
The proper second page of the results is returned starting from 'abc's position when the first page was fetched.

Scenario 2:
- fetch a page of followers. for example, the last follower in the page has a guid of 'abc'.
- before you fetch the next page, user `abc` un-follows you and then re-follows you which moves his position from last in page 1 to first of page 1.
- fetch the second page using 'abc' as the offsetId.
- haven't test this, but based on our discussion, I think the API will return page 1 again, albeit with user 'abc' in the first position instead of the last position.

Expected results:
The proper second page of the results is returned starting from 'abc's position when the first page was fetched.

@cpacia as per our discussion, perhaps a way to solve this is to have the offsetId reference a "follow" record as opposed to a GUID and, upon an un-follow, the record is logically deleted (i.e. status becomes deleted).

**Comment by cpacia** [comment]
I agree it's potentially an issue if an unfollow comes at the exact second you're scrolling throw your list, but I don't like the idea of the having the followers table grow infinitely with each follow and unfollow. 

Someone maliciously sends follows and unfollows it just grows your disk usage.

**Comment by rmisio** [comment]
It's not the exact second. As someone is scrolling through a list, they may be taking some time to read the list. Also, they may be distracted and start chatting with someone for a few minutes while and then continue scrolling.

It's a very real scenario that I don't think we can discount.

To prevent malicious abuse, we need to think of some other approach to solve that. One, which is necessary anyways, is to throttle the amount of un-follows / follows in a given window. Another, to help mitigate this is to every so often clean up "deleted" records that are older than a certain threshold (e.g. older than day or two - it's doubtful someone will be taking a break of more than a day during scrolling).

Anyhow, I guarantee this scenario will show up and there will be a bug when a duplicate page of followers is returned or an empty page is returned when the tab head reflects that more followers exist.

**Comment by cpacia** [comment]
You'll still have the issue if incorrect returns if someone if loading the followers while a garbage collection service is running.

**Comment by rmisio** [comment]
Not if the garbage collection is cleaning records older than a day or two. Then, yes, you'd have an issue if someone starts a scroll and then takes a day or two break and then continues the scroll, but the chances of that happening are about the same as me winning the power ball.

**Comment by rmisio** [comment]
There's an alternate way we could do this that wouldn't require "soft" deleting records and therefore growing the DB. Instead of paginating from an `offsetId`, we could make it timestamp based, e.g. `/ob/following?before=<timestamp> &limit=10` which would return the ten records with a createdAt timestamp older than the passed in timestamp.

If we went this route, it would require that the API return the createdAt timestamp for each record, instead of just the guid:

```
[
  {
    guid: 123,
    createdAt: <timestamp>
  },
  {
    guid: 124,
    createdAt: <timestamp>
  },
]
```

Also, in the `follow` socket, we would also need the timestamp provided as well.

**Comment by jjeffryes** [comment]
@cpacia and @rmisio is this issue still a problem?

**Comment by rmisio** [comment]
Yes, it is possible that the last loaded user on the client may no longer be in the list (e.g., has unfollowed) and if we fetch with that user as the offset ID, either an empty list will return or, if the user has refollowed, duplicate users will be returned from the top of the list.

I think it's a hole in the API that should be addressed, especially if it's not a huge refactor.

@cpacia What do you think of making the offset timestamp instead of ID based as I describe [here](https://github.com/OpenBazaar/openbazaar-go/issues/90#issuecomment-255161179)? I think that would solve this issue.

**Comment by rmisio** [comment]
> Do we want to remove the string ID here in exchange for the numberical ID like we did on notifications recently?

We could. But... that has nothing to do with this bug, so maybe there's somewhere else we could track that?

**Comment by placer14** [comment]
It would indirectly solve it. Because numerical IDs are sequential and the code would handle for the case where you have [3, 6, 9, 12] and ask for offsetId > 10.

**Comment by rmisio** [comment]
Oh, I think you're right because now it's using the peerId as the ID and that means it's essentially a reused ID if the user is unfollowed and then followed.

The numeric id would probably solve this (we would need the id to be returned in both the http and socket apis).
